src/routes/threlte/projectUtils.ts

```diff
@@ -0,0 +1,544 @@
+import { Vector2, Vector3, type Vector2Like } from "three"
+import type { Entity, ExtrusionHistoryStep, HistoryStep, Message, MessageHistory, PlaneHistoryStep, PointHistoryStep, Point2D, SketchHistoryStep, WithTarget, WorkBench } from "../../types"
+import type { Realization as WasmRealization } from "cadmium"
+import { isDeleteArcs, isDeleteCircles, isDeleteLines, isNewCircleBetweenPoints, isNewExtrusion, isNewLineOnSketch, isNewPointOnSketch2, isNewRectangleBetweenPoints, isNewSketchOnPlane, isRenameStep, isSetSketchPlane, isUpdateExtrusion } from "../../typeGuards"
+import { _isDevelopment } from "../+layout"
+
+// prettier-ignore
+const log = (function () { const context = "[projectUtils.ts]"; const color = "aqua"; return Function.prototype.bind.call(console.log, console, `%c${context}`, `font-weight:bold;color:${color};`) })()
+
+export const CIRCLE_TOLERANCE = 0.05
+
+export function isPoint(feature: HistoryStep): feature is PointHistoryStep {
+	return feature.data.type === "Point"
+}
+export function isPlane(feature: HistoryStep): feature is PlaneHistoryStep {
+	return feature.data.type === "Plane"
+}
+export function isExtrusion(feature: HistoryStep): feature is ExtrusionHistoryStep {
+	return feature.data.type === "Extrusion"
+}
+export function isSketch(feature: HistoryStep): feature is SketchHistoryStep {
+	return feature.data.type === "Sketch"
+}
+
+export function arraysEqual(a: any[], b: any[]) {
+	if (a.length !== b.length) return false
+	for (let i = 0; i < a.length; i++) {
+		if (a[i] !== b[i]) return false
+	}
+	return true
+}
+
+function sendWasmMessage(message: Message) {
+	let wp = get(wasmProject)
+	const messageStr = JSON.stringify(message)
+	log("[sendWasmMessage] sending message:", message)
+	let reply = wp.send_message(messageStr)
+	log("[sendWasmMessage] reply:", reply)
+	let result = JSON.parse(reply)
+
+	messageHistory.update((history: MessageHistory[]) => {
+		log("[sendWasmMessage] [messageHistory.update] update:", { message, result })
+		return [...history, { message, result }]
+	})
+	return result
+}
+
+export function updateExtrusion(extrusionId: string, sketchId: string, length: number, faceIds: string[]) {
+	const message: Message = {
+		UpdateExtrusion: {
+			workbench_id: get(workbenchIndex),
+			sketch_id: sketchId,
+			face_ids: faceIds.map(id => +id), // on browser side ids are strings - coerce face ids to numbers here to suit rust
+			length,
+			offset: 0.0,
+			extrusion_name: 'Extra',
+			direction: 'Normal',
+			extrusion_id: extrusionId
+		}
+	}
+	const isValid = checkWasmMessage(message)
+	const hasFaceIds = notEmpty(message.UpdateExtrusion.face_ids)
+	if (isValid) {
+		sendWasmMessage(message)
+		workbenchIsStale.set(true)
+		if(hasFaceIds) {
+			log("[updateExtrusion]", "[checkWasmMessage]", "is valid,", "sending message...", message)
+			// sendWasmMessage(message)
+		} else log("[updateExtrusion]", "[checkWasmMessage]", "is valid,", "but face_ids is empty," ,"NOT sending message:", message)
```

Comment:
I think it is valid for a user to create an extrusion, assign face IDs to it, then later come back and edit that extrusion, temporarily removing all the face IDs and assigning new ones. I think that this function would support that flow as written, so maybe this comment is out of date?

Reply:
Yes this is from before I understood how this part worked. Could be useful logging this for now while developing. 

As for sending the message, you're saying we should send the message whether or not face_ids is empty? So the current code is fine and I'll just remove the comments? 

Oh! I see what you mean, I'll remove "NOT sending message:" because we are sending it and that is indeed stale from when experimenting!

To be clear; you're happy with the code, just tidy up the logging messages and remove the comments?

Reply:
yes I'm happy with the code as is, was just hoping you could remove the comments that are stale!